src/vehicle.cpp

```diff
@@ -5704,190 +5705,180 @@ bool vehicle::shift_if_needed() {
+    if( one_in( explosion_chance ) ) {
+        g->u.add_memorial_log(pgettext("memorial_male","The fuel tank of the %s exploded!"),
+            pgettext("memorial_female", "The fuel tank of the %s exploded!"),
+            name.c_str());
+        g->explosion( global_part_pos3( p ), pow, 0.7, data.fiery_explosion );
+        parts[p].hp = 0;
+        parts[p].amount = 0;
+    }
+
+    return true;
+}
+
+int vehicle::damage_direct( int p, int dmg, damage_type type )
+{
+    if( parts[p].hp <= 0 ) {
+        return break_off( p, dmg );
+    }
+
+    int tsh = std::min( 20, part_info(p).durability / 10 );
+    if( dmg < tsh && type != DT_TRUE ) {
+        if( type == DT_HEAT && part_flag( p, "FUEL_TANK" ) ) {
+            explode_fuel( p, type );
         }
+
+        return dmg;
     }
-    if (dres < 0)
-        dres = 0;
-    return dres;
+
+    dmg -= part_info( p ).damage_reduction[ type ];
+    int dres = dmg - parts[p].hp;
+    int last_hp = parts[p].hp;
+    parts[p].hp -= dmg;
+    parts[p].hp = std::max( parts[p].hp, 0 );
+
+    if( parts[p].hp == 0 && last_hp > 0 ) {
+        insides_dirty = true;
+        pivot_dirty = true;
+    }
+
+    if( part_flag( p, "FUEL_TANK" ) ) {
+        explode_fuel( p, type );
+    } else if( parts[p].hp <= 0 && part_flag(p, "UNMOUNT_ON_DAMAGE") ) {
+        g->m.spawn_item( global_part_pos3( p ), part_info( p ).item, 1, 0, calendar::turn );
+        remove_part( p );
+    }
+
+    return std::max( dres, 0 );
 }
 
-void vehicle::leak_fuel (int p)
+void vehicle::leak_fuel( int p )
 {
-    if (!part_flag(p, "FUEL_TANK")) {
+    if( !part_flag( p, "FUEL_TANK" ) ) {
+        return;
+    }
+
+    const itype_id &ft = part_info( p ).fuel_type;
+
+    // @todo Move leak data to json
+    if( ft != fuel_type_gasoline && ft != fuel_type_diesel ) {
+        parts[p].amount = 0;
         return;
     }
 
-    const itype_id &ft = part_info(p).fuel_type;
     item leak( ft, calendar::turn );
 
-    if (ft == fuel_type_gasoline || ft == fuel_type_diesel) {
-        tripoint minp = global_pos3();
-        tripoint maxp = minp;
-        minp.x -= 2;
-        minp.y -= 2;
-        maxp.x += 2;
-        maxp.y += 2;
-
-        for ( const tripoint &pt : g->m.points_in_rectangle(minp, maxp) ){
-            if (g->m.passable(pt) && one_in(2)) {
-                int leak_amount = rng(79, 121);
-
-                if (parts[p].amount < leak_amount) {
-                    parts[p].amount = 0;
-                    return;
-                }
+    for( const tripoint &pt : g->m.points_in_radius( global_part_pos3( p ), 2 ) ) {
+        if( one_in( 2 ) && g->m.passable( pt ) ) {
+            int leak_amount = rng(79, 121);
 
-                leak.charges = leak_amount;
-                g->m.add_item_or_charges( pt, leak );
-                parts[p].amount -= leak_amount;
+            if( parts[p].amount < leak_amount ) {
+                parts[p].amount = 0;
+                return;
```

Comment:
This returns too early. The leak amount has been removed from the part, but has not yet been placed on the map. It is essentially gone. (Or was that intended?)